packages/@lwc/template-compiler/src/codegen/codegen.ts

```diff
@@ -578,7 +583,14 @@ export default class CodeGen {
         // Depth-first traversal. We assign a partId to each element, which is an integer based on traversal order.
         while (stack.length > 0) {
             const node = stack.shift()!;
-            partId++;
+
+            if (!(isComment(node) && !this.preserveComments)) {
+                // Skip adding part index for comments when preserveComments is disabled.
+                // The partId represents the nodes in the DOM tree at runtime, comments will be
+                // stripped out when preserveComments is disabled.
+                partId++;
+            }
```

Comment:
Exclude comments from the `partId` count when `lwc:preserve-comments` is disabled as they will be stripped out at runtime.